index.js

```diff
@@ -304,23 +302,50 @@ if (instrumentPromise) {
   var PromiseMethods = ['accept', 'all', 'defer', 'race', 'reject', 'resolve'];
 
   PromiseMethods.forEach(function(key) {
-    global.Promise[key] = Promise[key];
+    wrappedPromise[key] = Promise[key];
   });
-}
 
-function wrapThen(original) {
-  return function wrappedThen() {
-    var promise = this;
-    return original.apply(promise, Array.prototype.map.call(arguments, bind));
-
-    // wrap callbacks (success, error) so that the callbacks will be called as a
-    // continuations of the accept or reject call using the __asl__wrapper created above.
-    function bind(fn) {
-      if (typeof fn !== 'function') return fn;
-      return function(val) {
-        if (!promise.__asl_wrapper) return fn.call(this, val);
-        return promise.__asl_wrapper(this, fn, val);
-      };
+  global.Promise = wrappedPromise;
+
+  function ensureAslWrapper(promise) {
+    if (!promise.__asl_wrapper) {
+      promise.__asl_wrapper = wrapCallback(propagateAslWrapper);
+    }
+  }
+
+  function propagateAslWrapper(ctx, fn, result, next) {
+    var nextResult;
+    try {
+      nextResult = fn.call(ctx, result);
+      return nextResult;
+    } finally {
+      // Wrap any resulting futures as continuations.
+      if (nextResult instanceof wrappedPromise) {
```

Comment:
so this is kind of a dumb edge case, but I had a discussion earlier with a coworker about what this instanceof check should be checking against.  I believe it would be better to use the original promise incase you pass an unwrapped promise from to then. Your not going to get the best results if your mixing unwrapped promises into things, but for the sake of interoperability with other monkey patching I think its best to leave the most basic check possible here.